npm/podController.go

```diff
@@ -425,26 +407,30 @@ func (c *podController) syncAddedPod(podObj *corev1.Pod) error {
 // syncAddAndUpdatePod handles updating pod ip in its label's ipset.
 func (c *podController) syncAddAndUpdatePod(newPodObj *corev1.Pod) error {
 	var err error
-	ipsMgr := c.npMgr.NsMap[util.KubeAllNamespacesFlag].IpsMgr
-
-	// Create ipset related to namespace which this pod belong to if it does not exist.
 	newPodObjNs := util.GetNSNameWithPrefix(newPodObj.Namespace)
-	if _, exists := c.npMgr.NsMap[newPodObjNs]; !exists {
-		if err = ipsMgr.CreateSet(newPodObjNs, []string{util.IpsetNetHashFlag}); err != nil {
+
+	// lock before using nsMap since nsMap is shared with namespace controller
+	c.npmNamespaceCache.Lock()
+	if _, exists := c.npmNamespaceCache.nsMap[newPodObjNs]; !exists {
+		// Create ipset related to namespace which this pod belong to if it does not exist.
+		if err = c.ipsMgr.CreateSet(newPodObjNs, []string{util.IpsetNetHashFlag}); err != nil {
+			c.npmNamespaceCache.Unlock()
 			return fmt.Errorf("[syncAddAndUpdatePod] Error: failed to create ipset for namespace %s with err: %v", newPodObjNs, err)
 		}
 
-		if err = ipsMgr.AddToList(util.KubeAllNamespacesFlag, newPodObjNs); err != nil {
+		if err = c.ipsMgr.AddToList(util.KubeAllNamespacesFlag, newPodObjNs); err != nil {
```

Comment:
I did not see where we are creating KubeAllNamespacesFlag ipset, can you point me to that .. 
Same with kube-system ns ipset

Reply:
ignore: NPM creates ipset while adding to set